hook-janitor/src/webhooks.rs

```diff
@@ -1,55 +1,454 @@
+    #[sqlx(try_from = "i64")]
+    successes: u32,
+}
+
+#[derive(sqlx::FromRow, Debug)]
+struct FailedRow {
+    // App Metrics truncates/aggregates rows on the hour, so we take advantage of that to GROUP BY
+    // and aggregate to select fewer rows.
+    hour: DateTime<Utc>,
+    // A note about the `try_from`s: Postgres returns all of those types as `bigint` (i64), but
+    // we know their true sizes, and so we can convert them to the correct types here. If this
+    // ever fails then something has gone wrong.
+    #[sqlx(try_from = "i64")]
+    team_id: u32,
+    #[sqlx(try_from = "i64")]
+    plugin_config_id: u32,
+    #[sqlx(json)]
+    last_error: WebhookJobError,
+    #[sqlx(try_from = "i64")]
+    failures: u32,
 }
 
 impl WebhookCleaner {
     pub fn new(
         queue_name: &str,
         table_name: &str,
         database_url: &str,
-        batch_size: u32,
         kafka_producer: FutureProducer<KafkaContext>,
         app_metrics_topic: String,
-        plugin_log_entries_topic: String,
-    ) -> Result<Self, CleanerError> {
+    ) -> Result<Self> {
         let queue_name = queue_name.to_owned();
         let table_name = table_name.to_owned();
         let pg_pool = PgPoolOptions::new()
+            .acquire_timeout(Duration::from_secs(10))
             .connect_lazy(database_url)
-            .map_err(|error| CleanerError::PoolCreationError { error })?;
+            .map_err(|error| WebhookCleanerError::PoolCreationError { error })?;
 
         Ok(Self {
             queue_name,
             table_name,
             pg_pool,
-            batch_size,
             kafka_producer,
             app_metrics_topic,
-            plugin_log_entries_topic,
         })
     }
+
+    #[allow(dead_code)] // This is used in tests.
+    pub fn new_from_pool(
+        queue_name: &str,
+        table_name: &str,
+        pg_pool: PgPool,
+        kafka_producer: FutureProducer<KafkaContext>,
+        app_metrics_topic: String,
+    ) -> Result<Self> {
+        let queue_name = queue_name.to_owned();
+        let table_name = table_name.to_owned();
+
+        Ok(Self {
+            queue_name,
+            table_name,
+            pg_pool,
+            kafka_producer,
+            app_metrics_topic,
+        })
+    }
+
+    async fn start_serializable_txn(&self) -> Result<Transaction<'_, Postgres>> {
+        let mut tx = self
+            .pg_pool
+            .begin()
+            .await
+            .map_err(|e| WebhookCleanerError::StartTxnError { error: e })?;
+
+        // We use serializable isolation so that we observe a snapshot of the DB at the time we
+        // start the cleanup process. This prevents us from accidentally deleting rows that are
+        // added (or become 'completed' or 'failed') after we start the cleanup process.
+        //
+        // If we find that this has a significant performance impact, we could instead move
+        // rows to a temporary table for processing and then deletion.
```

Comment:
For normal webhook processing, we are skipping all the locked rows (for two reasons, we run with `SKIP LOCKED` and we check for status not in (`failed`, `completed`). So, I don't expect major performance impact there. If this transaction stays open for too long, that could delay auto-vacuum, and eventually degrade overall performance. However, I don't see this running for that long, and as long as we timeout effectively on Postgres side we should catch any lingering open transactions.